pkg/frontend/querymiddleware/astmapper/subquery_spin_off.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+const (
+	SubqueryMetricName      = "__subquery_spinoff__"
+	SubqueryQueryLabelName  = "__query__"
+	SubqueryRangeLabelName  = "__range__"
+	SubqueryStepLabelName   = "__step__"
+	SubqueryOffsetLabelName = "__offset__"
+
+	DownstreamQueryMetricName = "__downstream_query__"
+	DownstreamQueryLabelName  = "__query__"
+)
+
+type subquerySpinOffMapper struct {
+	ctx             context.Context
+	defaultStepFunc func(rangeMillis int64) int64
+
+	logger log.Logger
+	stats  *SubquerySpinOffMapperStats
+}
+
+// NewSubquerySpinOffMapper creates a new instant query mapper.
+func NewSubquerySpinOffMapper(ctx context.Context, defaultStepFunc func(rangeMillis int64) int64, logger log.Logger, stats *SubquerySpinOffMapperStats) ASTMapper {
+	queryMapper := NewASTExprMapper(
+		&subquerySpinOffMapper{
+			ctx:             ctx,
+			defaultStepFunc: defaultStepFunc,
+			logger:          logger,
+			stats:           stats,
+		},
+	)
+
+	return NewMultiMapper(
+		queryMapper,
+	)
+}
+
+// MapExpr implements the ASTMapper interface.
+// The strategy here is to look for aggregated subqueries (all subqueries should be aggregated) and spin them off into separate queries.
+// The frontend does not have internal control of the engine,
+// so MapExpr has to remap subqueries into "fake metrics" that can be queried by a Queryable that we can inject into the engine.
+// This "fake metric selector" is the "__subquery_spinoff__" metric.
+// For everything else, we have to pass it through to the downstream execution path (other instant middlewares),
+// so we remap them into a "__downstream_query__" selector.
+//
+// See sharding.go and embedded.go for another example of mapping into a fake metric selector.
+func (m *subquerySpinOffMapper) MapExpr(expr parser.Expr) (mapped parser.Expr, finished bool, err error) {
+	if err := m.ctx.Err(); err != nil {
+		return nil, false, err
+	}
+
+	// Immediately clone the expr to avoid mutating the original
+	expr, err = cloneExpr(expr)
+	if err != nil {
+		return nil, false, err
+	}
+
+	downstreamQuery := func(expr parser.Expr) (mapped parser.Expr, finished bool, err error) {
+		if countSelectors(expr) == 0 {
+			return expr, false, nil
+		}
+		selector := &parser.VectorSelector{
+			Name: DownstreamQueryMetricName,
+			LabelMatchers: []*labels.Matcher{
+				labels.MustNewMatcher(labels.MatchEqual, DownstreamQueryLabelName, expr.String()),
+			},
+		}
+		m.stats.AddDownstreamQuery()
+		return selector, false, nil
+	}
+
+	switch e := expr.(type) {
+	case *parser.Call:
+		if len(e.Args) == 0 {
+			return expr, false, nil
+		}
+		lastArgIdx := len(e.Args) - 1
+		// The last argument will typically contain the subquery in an aggregation function
+		// Examples: last_over_time(<subquery>[5m:]) or quantile_over_time(0.5, <subquery>[5m:])
+		if sq, ok := e.Args[lastArgIdx].(*parser.SubqueryExpr); ok {
+			// @ is not supported
```

Comment:
[not blocking] I don't think it would be too complex to support `@` - it effectively changes the end timestamp used in the query range calculations.

Reply:
I haven't seen it used much overall. I'll keep it out for now